Fix issue #46

JaniruTEC opened the pull request:
Added MAPPING.md as common guideline for type-conversions
Added new Annotation (Unsigned) used to indicate that an integer type or integer type pointer refers to an unsigned value.
Corrected types of fields in structures to match Dokany
Added Unsigned-Annotation to all accessors of unsigned values
Added UnsignedNumbers class with conversion methods between unsigned integer types and String

Merging this PR closes issue #46

JaniruTEC committed:
Added MAPPING.md as common guidelines for type-conversion

JaniruTEC committed:
Updated MAPPING.md with references and updated Data Types

JaniruTEC committed:
Added Annotation (Unsigned) and fixed bugs

Added new Annotation (Unsigned) used to indicate that an integer type or integer type pointer refers to an unsigned value.
Used Unsigned-Annotation to mark parameter length in getDokanControlList method in DokanControl and all calls that lead to it
Changed getDokanControlList to use an unsigned int instead of a long (matching Dokany that uses ULONG) and finished method (removed TODO)
Changed length of array MountPoint to 260 to match Dokany

JaniruTEC committed:
Updated DokanControl

Updated DokanControl to use WinNT.MAX_PATH and rewrote getDokanControlList

JaniruTEC committed:
Corrected fields in structures and fixed wrong types

Corrected types of fields in structures to match Dokany
FIxed wrong types/usage in Constructor in DokanControl
FIxed wrong parameters in Constructor in DokanOptions
Fixed wrong parameters and added Unsigned-Annotation in accessors of the constructer

JaniruTEC committed:
Added Unsigned-Annotation to matching fields in all structures

JaniruTEC committed:
Updated MAPPING.md to use correct version of JNA (5.5.0)

JaniruTEC committed:
Added UnsignedNumbers utility class and marked values as unsigned

Added Unsigned-Annotation to all accessors of unsigned value
Added UnsignedNumbers class with conversion methods between unsigned integer types and String
Fixed toString method of DokanFileInfo and DokanOptions

JaniruTEC committed:
Merge remote-tracking branch 'upstream/develop' into feature/fix-46

infeo commented on src/main/java/dev/dokan/dokan_java/examples/DirListingFileSystem.java:
```diff
@@ -93,7 +94,7 @@ public int zwCreateFile(WString rawPath, DokanIOSecurityContext securityContext,
         }
 
-        long val = this.handleHandler.incrementAndGet();
+        @Unsigned long val = this.handleHandler.incrementAndGet();
```
Here the annotation is actually not needed, since this is "just" an example filesystem implementation.

JaniruTEC replied:
Objection: In my opinion the example filesystem is the best place to teach usage of the annotation.

infeo replied:
From my point of view the annotation is mostly to indicate how to treat a number given by or returned to the library. A developer using the filesystem can acknowledge it and handle it, but from the point in time where their code is executed it does not matter anymore, as long as they considered it when returning into the library again.

JaniruTEC replied:
In this case the sole purpose of the defined vars is to be returned to the filesystem. Therefore flagging them as "special" seems appropriate to me. The DirListingFileSystem should teach the correct way of using the dokan-java API and the best way to use it is to consistently keep track of unsigned values.

infeo commented on src/main/java/dev/dokan/dokan_java/structure/DokanControl.java:
```diff
@@ -19,12 +22,13 @@
      * Mount point. Can be "M:\" (drive letter) or "C:\mount\dokan" (path in NTFS)
      */
-    public char[] MountPoint = new char[256];
+    public char[] MountPoint = new char[WinNT.MAX_PATH];
```
Actually not part of this PR. But let's take it in.

infeo commented on src/main/java/dev/dokan/dokan_java/structure/DokanControl.java:
```diff
@@ -79,19 +83,28 @@ public DokanControl(Pointer p, long offset) {
-            list.add(new DokanControl(start));
-            for (int i = 1; i < length; i++) {
-                list.add(new DokanControl(start, i * list.get(0).size()));
+        /*
+         * Let's do the math:
+         * A list that uses an unsigned int (32 bit) as index could save up to 2^32 objects.
+         * Even if it only saved one byte in each entry (not accounting for the actual overhead of the entries themselves),
+         * that would be 2^32 bytes = 4 GB for that list alone!
+         * If the list of active Dokan MountPoints exceeds 2^31 (signed int) entries, we probably have other problems.
+         *
+         * But let's assume that this could still happen:
+         * In this case we want the application to crash so that someone else can fix that nonsense.
+         * "assert" seems like a good choice for this case (for once).
+         */
```
As discussed: Less sassy, and move it the javadocs part for documentation purpose.

JaniruTEC replied:
Resolved by commit 279b5b890b4b148a86fe25ad1d9b43f9fd7d8507

infeo commented on src/main/java/dev/dokan/dokan_java/UnsignedNumbers.java:
```diff
@@ -0,0 +1,22 @@
+    public static String toUnsignedString(@Unsigned long value) {
+        return Long.toUnsignedString(value);
+    }
+
```
As discussed: It would be nice to have also delegate methods for unsigned division and other operations, which are different from their signed counterparts.

JaniruTEC replied:
Resolved by commits 8a15e47 and 236347d

infeo commented on src/main/java/dev/dokan/dokan_java/Unsigned.java:
```diff
@@ -0,0 +1,17 @@
+import static java.lang.annotation.ElementType.*;
+
+
+@Documented
```
Please add some explaining JavaDoc. The following questions should be answered:

1. What does this annotation mean.
1. Why is this annotation needed.
1. Are there pitfalls (e.g. division of unsigned numbers)

JaniruTEC replied:
Revolved by commit b9aa49a902581c5ad01f01aeed07096d889973b0

infeo commented on MAPPING.md:
```diff
@@ -0,0 +1,12 @@
+# Overview
+| C Data Type         | Windows Data Type - Macro | Size (Bytes/bits) | java/JNA Data Type                |
+|---------------------|---------------------------|-------------------|-----------------------------------|
+| char/byte           | UCHAR                     | 1B/8b             | byte                              |
+| unsigned short      | USHORT                    | 2B/16b            | short (@Unsigned)/WinDef.USHORT   |
+| unsigned long (int) | ULONG                     | 4B/32b            | int (@Unsigned)/WinDef.ULONG      |
+| unsigned long (int) | DWORD/AccessMask          | 4B/32b            | int (@Unsigned)/WinDef.DWORD      |
+| unsigned __int64    | ULONG64                   | 8B/64b            | long (@Unsigned)/WinDef.ULONGLONG |
+# References
+Basic Types/Size: [Web](https://docs.microsoft.com/en-us/cpp/c-language/storage-of-basic-types?view=vs-2019) | [Permalink](https://web.archive.org/web/20200616151823/https://docs.microsoft.com/en-us/cpp/c-language/storage-of-basic-types?view=vs-2019) <br>
+Windows Data Types/Mapped C Data Types: [Web](https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types) | [Permalink](https://web.archive.org/web/20200616152122/https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types) <br>
+JNA Data Types: [Web](https://java-native-access.github.io/jna/5.5.0/javadoc/com/sun/jna/platform/win32/WinDef.html) | [Permalink](https://web.archive.org/web/20200616212719/https://java-native-access.github.io/jna/5.5.0/javadoc/com/sun/jna/platform/win32/WinDef.html) <br>
```
Imho not our responsiblilty to document this, because this is only interesting for developers of this library. But since the primary source is linked, it can stay.

JaniruTEC replied:
See: https://github.com/dokan-dev/dokan-java/pull/48#issuecomment-647735932

infeo reviewed (changes requested):
Generally I approve, but the (already previously) discussed changes need to be implemented.

JaniruTEC committed:
Moved my comment and took out the sass

Moved my comment and took out the sass to make it more professional

See: https://github.com/dokan-dev/dokan-java/pull/48#discussion_r443744844

JaniruTEC committed:
Replaced usage of "assert" with explicit check

Replaced usage of "assert" with explicit check to make sure that the check is done, even if assertions are disabled.

From the java language guide (linked below):
"Do not use assertions for argument checking in public methods.
Argument checking is typically part of the published specifications (or contract) of a method, and these specifications must be obeyed whether assertions are enabled or disabled."

The next sentence "Another problem with using assertions for argument checking is that erroneous arguments should result in an appropriate runtime exception (such as IllegalArgumentException, IndexOutOfBoundsException, or NullPointerException). An assertion failure will not throw an appropriate exception." doesn't apply in this case as this method isn't part of the API. It's only used internally and this check is implemented solely for the purpose of making sure everything in this library is working fine.

See: https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html or https://web.archive.org/web/20200622193110/https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html

JaniruTEC commented:
To document the reasoning for creating [MAPPING.md:](https://github.com/dokan-dev/dokan-java/blob/5895e92ead7046468f4dc4732cfb8c758d530089/MAPPING.md)
1. It's important to use the same mapping all over the project, therefore establishing a guideline seemed to be important.
2. It gives easy access to any sources regarding this matter.
3. It helps users of this API, as well as new developers (as well as old ones after a few weeks have passed) to understand the inner workings of the conversions between Dokany and dokan-java.

JaniruTEC committed:
Added additional methods to UnsignedNumbers

Added methods to compare, divide and get the remainder of unsigned numbers to UnsignedNumbers.

See: https://github.com/dokan-dev/dokan-java/pull/48#discussion_r443746630

JaniruTEC committed:
Added methods for unsigned conversion between integer types

Added methods for unsigned conversion between integer types

JaniruTEC committed:
Added missing constants from last commit to UnsignedNumbers

Added constants that were missing from my last commit to UnsignedNumbers to allow compilation/fix compiler errors. Sorry!

JaniruTEC committed:
Changed @ Unsiged to only use TYPE_PARAMETER and TYPE_USE as targets

Removed all unnecessary values from the @ Targets of @ Unsigned

infeo commented on src/main/java/dev/dokan/dokan_java/structure/DokanControl.java:
```diff
@@ -73,25 +77,33 @@ public DokanControl(Pointer p, long offset) {
 
     /**
-     * Creates a java {@link List} of {@link DokanControl} strcutures given the pointer returned by NativeMethods#DokanGetMountPointList(boolean, LongByReference).
+     * Creates a java {@link List} of {@link DokanControl} strcutures given the pointer returned by NativeMethods#DokanGetMountPointList(boolean, LongByReference).<br>
```
typo.

JaniruTEC replied:
Resolved by commit 75e582c

infeo reviewed (dismissed):
Add the Javadoc for the annotation and this PR is ready to be merged.

JaniruTEC committed:
Fixed typos in package dev.dokan.dokan_java.structure.*

See: https://github.com/dokan-dev/dokan-java/pull/48#discussion_r446683030

JaniruTEC committed:
Added JavaDoc for @ Unsigned

See: https://github.com/dokan-dev/dokan-java/pull/48#discussion_r443747818

JaniruTEC commented on src/main/java/dev/dokan/dokan_java/Unsigned.java:
```diff
@@ -10,6 +10,182 @@
+ * and <a href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.11">&sect4.11</a> of "The Java&reg Language Specification" for
+ * further reference.<br>
+ * <br>
+ * At least the following cases ("minimum usage") should be annotated with <i>@Unsigned</i> to
```
In the future it may be useful to check our/foreign source code for compliance to this rule, e.g. with [the Checker Framework.](https://github.com/typetools/checker-framework)

JaniruTEC replied:
// Do not close this conversation //

JaniruTEC commented on src/main/java/dev/dokan/dokan_java/Unsigned.java:
```diff
@@ -0,0 +1,196 @@
+ * and <a href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.11">&sect4.11</a> of "The Java&reg Language Specification" for
+ * further reference.<br>
+ * <br>
+ * At least the following cases ("minimum usage") should be annotated with <i>@Unsigned</i> to
+ * guarantee the best quality of code:
```
In the future it may be useful to check our/foreign source code for compliance to this rule, e.g. with [the Checker Framework.](https://github.com/typetools/checker-framework)

JaniruTEC replied:
// Do not resolve or close. //
